akka-http/server/src/main/scala/endpoints4s/akkahttp/server/BasicAuthentication.scala

```diff
@@ -27,33 +27,8 @@ trait BasicAuthentication extends algebra.BasicAuthentication with EndpointsWith
       tuplerHCred: Tupler.Aux[H, Credentials, HCred],
       tuplerUEHCred: Tupler.Aux[UE, HCred, Out]
   ): Request[Out] = {
-    val (m, u, e, h) = (method, url, entity, headers)
     new Request[Out] {
-      type UrlData = U
-      type HeadersData = (H, Option[Credentials])
-      type EntityData = E
-
-      def url: Url[UrlData] = u
-      def method: Method = m
-      def entity: RequestEntity[EntityData] = e
-      def headers: RequestHeaders[HeadersData] = h ++ authHeader
-
-      private[server] def aggregateAndValidate(
-          urlData: UrlData,
-          entityData: EntityData,
-          headersData: HeadersData
-      ): Validated[Out] =
-        headersData match {
-          case (_, None) =>
-            // Note: in practice that should not happen because the method `aggregateAndValidate` is
-            // only called from the final method `matches`, if `matchAndParseHeaders` succeeded.
-            // However, here we override `matchAndParseHeaders` to fail in case the credentials are missing.
-            sys.error(
-              "This request transformation is currently unsupported. You can't transform further an authenticated request."
-            )
-          case (h, Some(credentials)) =>
-            Valid(tuplerUEHCred(tuplerUE(urlData, entityData), tuplerHCred(h, credentials)))
-        }
+      type UrlAndHeaders = (U, H, Credentials)
```

Comment:
It is nice to see so much boilerplate go, though if I understand the changes correctly we do lose the ability to get the individual parts of a request in the interpreters. I would need to see in our own implementation of the auth to see whether that will give problems.

Reply:
That’s true, we might lose some power, but I believe in practice that should not be an issue (ie, we used to have “too much power”). But I might be wrong, so please let me know more precisely what are your needs.

The way it works in the use cases that are part of endpoints4s (`mapRequest` method and authentication support) is that we can always access the content of the headers, URL, and entity by calling `someRequest.matchAndParseUrlAndHeadersDirective` or `someRequest.parseEntityDirective`, which return the information parsed from the URL, headers, and entity.

Reply:
Yes, I have looked at our implementation of auth based on the current design and I believe the changes in this PR would still make that possible.

Reply:
@bmeesters Happy to know that the changes made in this PR still support your use case. Do you think it makes things simpler or more complex, though?

Reply:
TBH I don't think it will make much difference. We did need to copy over some package private methods that I suppose are not needed anymore, so that is nice. But our implementation doesn't lean on the types and `url`, `entity` etc. methods of the interpreter `Request`. We instead override the apply method in the `Endpoint` instead to some auth there.

Also for more authorization use case we instead need to do `myEndpoint.handleBearerAuthentication.implementedBy` to ensure we have access in the server to user information after authentication is done. As far as I understood it was very hard to do in the algebra because this type of information is only necessary in the server, and it is in between the request (not an input to the endpoint, but only after handing the request). But I believe the middlewares could make this redundant.

That said, @agboom has been the one implementing auth in our internal library on top of endpoints4s, so he knows all the details that I might have missed.

Reply:
> Also for more authorization use case we instead need to do `myEndpoint.handleBearerAuthentication.implementedBy` to ensure we have access in the server to user information after authentication is done. As far as I understood it was very hard to do in the algebra because this type of information is only necessary in the server, and it is in between the request (not an input to the endpoint, but only after handing the request). But I believe the middlewares could make this redundant.

Yes, maybe, but I am not so sure of that. I would be interested in knowing more about this use case to validate the approach taken here https://github.com/julienrf/endpoints/pull/2.

Reply:
I'm working on a barebones Authentication algebra based on our internal library, but adapted to your middlewares approach in https://github.com/julienrf/endpoints/pull/2. After I'm further along I'll share it somewhere in the coming week.

Reply:
Hi @julienrf, sorry for my absence. I did make progress on the barebones algebra I mentioned above, but not as much as I wanted. I can push something this evening that will give an idea of my intended use-case.

Reply:
No need to hurry :)